Apply dynamic-range-limit to HDR WebGPU canvas

squelart opened the pull request:
#### 959ddfba001700372b372d86ca7ef0b30574c604
<pre>
Apply dynamic-range-limit to HDR WebGPU canvas
<a href="https://bugs.webkit.org/show_bug.cgi?id=294831">https://bugs.webkit.org/show_bug.cgi?id=294831</a>
<a href="https://rdar.apple.com/148558614">rdar://148558614</a>

Reviewed by Mike Wyrzykowski.

Handle both dynamic-range-limit and suppress-HDR, to apply whichever
constrains the most.

* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::createContextWebGPU):
(WebCore::HTMLCanvasElement::dynamicRangeLimitDidChange):
Stores the current dynamic-range-limit, and sets the effective limit
(which accounts for the suppress-HDR state) in the rendering context.

(WebCore::HTMLCanvasElement::getContextEffectiveDynamicRangeLimitValue const):
For testing.

* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::setDynamicRangeLimit):
Default dynamic-range-limit handler, does nothing.

(WebCore::CanvasRenderingContext::getEffectiveDynamicRangeLimitValue const):
For testing.

* Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContextCocoa::GPUCanvasContextCocoa):
Observer now also passes suppressHDR to updateScreenHeadroom().

(WebCore::interpolateHeadroom):
(WebCore::GPUCanvasContextCocoa::computeContentsHeadroom):
Computes the headroom based on the screen headroom (which may already
be suppressed), and the dynamic-range-limit.

(WebCore::GPUCanvasContextCocoa::updateContentsHeadroom):
Computes the headroom and updates the compositorIntegration with it.

(WebCore::GPUCanvasContextCocoa::updateScreenHeadroom):
Called by observer, stores current headroom and suppressHDR, and
calls updateContentsHeadroom.

(WebCore::GPUCanvasContextCocoa::updateScreenHeadroomFromScreenProperties):
Fetches suppressHDR and current headroom from all screen properties, and
calls updateContentsHeadroom.

(WebCore::GPUCanvasContextCocoa::setDynamicRangeLimit):
WebGPU dynamic-range-limit handler, stores the new limits and:
- If there&apos;s no current screen data, calls
  updateScreenHeadroomFromScreenProperties, because dynamic-range-limit
  needs the current headroom and suppres-HDR flag to work,
- Otherwise just calls updateContentsHeadroom.

(WebCore::GPUCanvasContextCocoa::getEffectiveDynamicRangeLimitValue const):
For testing, combines dynamic-range-limit with suppress-HDR.

(WebCore::GPUCanvasContextCocoa::surfaceBufferToImageBuffer):
Refactored into updateScreenHeadroomFromScreenProperties.

(WebCore::GPUCanvasContextCocoa::configure):
Resets both current headroom and suppressHDR.

* Source/WebCore/rendering/RenderHTMLCanvas.h:
* Source/WebCore/rendering/RenderHTMLCanvas.cpp:
(WebCore::RenderHTMLCanvas::styleDidChange):
Catches dynamic-range-limit changes and notifies element.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::changeRequiresLayerRepaint const):
Ensures that dynamic-range-limit changes repaint the element, as its
effective colors may change.

* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::getContextEffectiveDynamicRangeLimitValue):
For testing: `internals.getContextEffectiveDynamicRangeLimitValue(canvas)`.

* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::screenPropertiesChanged):
Reworked suppress-headroom computation: Always 1.6, or current headroom
if lower.

* LayoutTests/http/tests/webgpu/webgpu/web_platform/canvas/dynamic-range-limit-constrained-disabled-expected.txt: Added.
* LayoutTests/http/tests/webgpu/webgpu/web_platform/canvas/dynamic-range-limit-constrained-disabled.html: Added.
* LayoutTests/http/tests/webgpu/webgpu/web_platform/canvas/dynamic-range-limit-expected.txt: Added.
* LayoutTests/http/tests/webgpu/webgpu/web_platform/canvas/dynamic-range-limit.html: Added.

Canonical link: <a href="https://commits.webkit.org/297105@main">https://commits.webkit.org/297105@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b5de0fab9485b576c16114b0a5a061279a05c00

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110542 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30201 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20634 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116568 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60809 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112505 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30880 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38789 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84061 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113490 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24653 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99537 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64502 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24014 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17677 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60363 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94036 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17736 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119358 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37582 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27900 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93027 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37956 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95808 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92850 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23661 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37856 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15597 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33555 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37477 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42948 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37140 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40479 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38848 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
```diff
@@ -814,6 +815,21 @@ - (void)setOwner:(WebCore::PlatformCALayer*)owner
+
+    m_dynamicRangeLimit = dynamicRangeLimit;
+
+    setLayerDynamicRangeLimit(m_layer.get(), dynamicRangeLimit);
```
I don't think using setLayerDynamicRangeLimit results in the intended behavior and it also does not allow us to correctly implement https://www.w3.org/TR/css-color-hdr-1/#dynamic-range-limit-mix

squelart replied:
For the 2nd point, I think this is moot: `dynamic-range-limit-mix()` is currently disabled anyway, because `[CALayer setPreferredDynamicRange:]` (called by `setLayerDynamicRangeLimit`) doesn't support arbitrary values. Once `CALayer` supports these, `mix` can be implemented.

I'm not sure why you think `setLayerDynamicRangeLimit` and therefore `setPreferredDynamicRange` doesn't have the intended behavior? Maybe we should discuss more... (Later this week, I won't be available much until then.)

mwyrzykowski replied:
Unless we can resolve the issue where SDR white becomes gray, I don't think we can use it. It is very noticeable in content and does not look correct

squelart replied:
But please remember that `dynamic-range-limit` is an author's choice (unlike OS-imposed dimming/suppressing), so the same author that chooses to use `standard` should be able to see the effects and deal with them, for example by having their WebGPU code generate pixel values that are appropriate to the SDR case.

squelart replied:
I've now removed the use of setPreferredDynamicRange, instead going for m_compositorIntegration->updateContentsHeadroom() as you had done for suppress-HDR notifications.

mattwoodrow commented on Source/WebCore/Modules/WebGPU/Implementation/WebGPUCompositorIntegrationImpl.cpp:
```diff
@@ -62,8 +62,10 @@ void CompositorIntegrationImpl::prepareForDisplay(uint32_t frameIndex, Completio
 #if HAVE(SUPPORT_HDR_DISPLAY)
-    for (auto& ioSurface : m_renderBuffers)
+    for (auto& ioSurface : m_renderBuffers) {
+        ALWAYS_LOG_WITH_STREAM(stream << "__GS__ CompositorIntegrationImpl[" << this << "]::updateContentsHeadroom(" << headroom << ") -> ioSurface[" << &ioSurface << "]->setContentEDRHeadroom(" << headroom << ")");
```
There's an 'HDR' log channel in WebCore now, if you want to make some of this logging more-permanent.

squelart replied:
Thanks for letting me know. But this is just temporary logging while we're debugging; I'll remove them all.

mattwoodrow commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -559,10 +560,14 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
 #endif
     }
 
+    if (properties.changedProperties & LayerChange::DynamicRangeLimitChanged)
```
Is this going to set `CADynamicRangeHigh` on basically every layer, even when we only have SDR.

Is there any downside to that? Should we we tracking the presence of HDR somewhere and only setting this if it's relevant?

squelart replied:
Good question! But to be considered HDR, a layer must have multiple attributes, including:
1. Explicit opt-in through non-standard `preferredDynamicRange` (used to be `wantsExtendedDynamicRange`),
2. Extended color space,
3. Floating-point color values.

2 and 3 are only activated when the WebGPU configuration explicitly sets the tone mapping mode to "extended".
So I think it's quite harmless to set the `preferredDynamicRange` alone.

Theoretically, we could also peek at the pixel values in the canvas, and only engage HDR if *any* color component is above 1.0, but that would probably way too much work, and would need continual monitoring when the content changes.

(And note that this draft may change direction and not use `preferredDynamicRange` in the end anyway; to be continued!)

squelart replied:
I've removed that code now.

mwyrzykowski commented on Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
```diff
@@ -190,15 +188,79 @@ static GPUIntegerCoordinate getCanvasHeight(const GPUCanvasContext::CanvasType&
+{
+    auto headroom = m_currentEDRHeadroom;
+    if (headroom >= 1.f && m_dynamicRangeLimit != PlatformDynamicRangeLimit::noLimit()) {
+        // FIXME: 1.0 doesn't actually constrain!?
```
Yes I think it reverts to some other behavior with 1.0 headroom

mwyrzykowski reviewed (approved):
Looks good with failing tests corrected, thank you!

mattwoodrow commented on Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
```diff
@@ -190,15 +187,93 @@ static GPUIntegerCoordinate getCanvasHeight(const GPUCanvasContext::CanvasType&
+{
+    m_suppressEDR = false;
+    m_currentEDRHeadroom = 1.f;
+    for (const auto& screenData : WebCore::getScreenProperties().screenDataMap.values()) {
```
Why are we iterating all the screens here? Shouldn't we know which screen this context will draw on, and use the values for that directly?

mwyrzykowski replied:
I think it's only used in the OffscreenCanvas path. Do we know which screen we will draw an OffscreenCanvas to?

But it is a bit confusing, perhaps we can change the name to make that obvious? Or inline the code into the call-site

mwyrzykowski replied:
oh nevermind, I see `|| m_currentEDRHeadroom < 1.f` in `GPUCanvasContextCocoa::setDynamicRangeLimit`

mattwoodrow replied:
OffscreenCanvas is indeed an interesting case.

If we're connected to an in-DOM canvas via a PlaceholderRenderingContext, then the answer is at least knowable, though potentially racy.

If the page is instead using a standalone OffscreenCanvas (and posting frames manually using ImageBitmaps) then I have no idea what happens. Do we even support HDR for those?

mattwoodrow replied:
I do wonder if this suggests that trying to tag surfaces in the context is the wrong approach, and we should instead be tagging them as they get presented (likely via setting `contentsHeadroom` on the CALayer instead).

That's a bigger change though, probably out of scope for this PR.

mattwoodrow commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1261,13 +1261,11 @@ void setLockdownModeEnabledGloballyForTesting(std::optional<bool> enabledForTest
+            constexpr auto maxSuppressedHeadroom = 1.6f;
+            auto suppressedHeadroom = std::min(maxSuppressedHeadroom, properties.currentEDRHeadroom);
+            properties.currentEDRHeadroom = suppressedHeadroom;
+            properties.maxEDRHeadroom = suppressedHeadroom;
```
Doesn't matter, since this value isn't used for anything, but I don't think we should be mutating the max here. I think we should leave it as a constant value, as the potential of the screen.

mattwoodrow commented on Source/WebCore/html/HTMLCanvasElement.cpp:
```diff
@@ -1013,6 +1016,25 @@ void HTMLCanvasElement::prepareForDisplay()
+#endif // ENABLE(PIXEL_FORMAT_RGBA16F)
+}
+
+double HTMLCanvasElement::effectiveDynamicRangeLimitValue() const
```
A 'ForTesting' suffix for these functions would be nice, since otherwise the fallback `-2.0` is somewhat confusing.

mwyrzykowski replied:
Agree, is it problematic to return 0 in the fallback?

squelart replied:
To avoid any trouble: I'll rename the function getContextEffectiveDynamicRangeValue, change the return to `optional<double>`, and the internals will throw an exception if there's no value.

mwyrzykowski commented on Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
```diff
@@ -190,15 +187,93 @@ static GPUIntegerCoordinate getCanvasHeight(const GPUCanvasContext::CanvasType&
+    static_assert(inLow < inHigh);
+    if (headroomForHigh <= headroomForLow)
+        return headroomForHigh;
+    return (headroomForHigh - headroomForLow) * (in - inLow) / (inHigh - inLow) + headroomForLow;
```
you could use `std::lerp` with `t = (in - inLow) / (inHigh - inLow)`

mwyrzykowski reviewed (changes requested):
Agree with @mattwoodrow at least for non-offscreen canvas, `GPUCanvasContextCocoa::updateScreenHeadroomFromScreenProperties` should never be called

mwyrzykowski commented on Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
```diff
@@ -111,9 +121,13 @@ class GPUCanvasContextCocoa final : public GPUCanvasContext {
     GPUIntegerCoordinate m_height { 0 };
-    float m_contentsHeadroom { 0.f };
+#if HAVE(SUPPORT_HDR_DISPLAY)
+    bool m_suppressEDR { false };
```
Very minor but can we move this with the other `bool` please (`m_compositingResultsNeedsUpdating`) to avoid unnecessary padding between member variables?

mwyrzykowski commented on Source/WebCore/html/canvas/CanvasRenderingContext.h:
```diff
@@ -122,7 +122,9 @@ class CanvasRenderingContext : public ScriptWrappable, public CanMakeWeakPtr<Can
     bool isHDR() const { return pixelFormat() == ImageBufferPixelFormat::RGBA16F; }
+    virtual void setDynamicRangeLimit(PlatformDynamicRangeLimit) { };
 #endif
+    virtual double effectiveDynamicRangeLimitValue() const { return -1.0; };
```
is it problematic to return either 0 or 1 instead of a negative dynamic range limit?

squelart replied:
Changed to return type `optional<double>`, `std::nullopt` in this case.

squelart committed:
Apply dynamic-range-limit to HDR WebGPU canvas
https://bugs.webkit.org/show_bug.cgi?id=294831
rdar://148558614

Reviewed by Mike Wyrzykowski.

Handle both dynamic-range-limit and suppress-HDR, to apply whichever
constrains the most.

* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::createContextWebGPU):
(WebCore::HTMLCanvasElement::dynamicRangeLimitDidChange):
Stores the current dynamic-range-limit, and sets the effective limit
(which accounts for the suppress-HDR state) in the rendering context.

(WebCore::HTMLCanvasElement::getContextEffectiveDynamicRangeLimitValue const):
For testing.

* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::setDynamicRangeLimit):
Default dynamic-range-limit handler, does nothing.

(WebCore::CanvasRenderingContext::getEffectiveDynamicRangeLimitValue const):
For testing.

* Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContextCocoa::GPUCanvasContextCocoa):
Observer now also passes suppressHDR to updateScreenHeadroom().

(WebCore::interpolateHeadroom):
(WebCore::GPUCanvasContextCocoa::computeContentsHeadroom):
Computes the headroom based on the screen headroom (which may already
be suppressed), and the dynamic-range-limit.

(WebCore::GPUCanvasContextCocoa::updateContentsHeadroom):
Computes the headroom and updates the compositorIntegration with it.

(WebCore::GPUCanvasContextCocoa::updateScreenHeadroom):
Called by observer, stores current headroom and suppressHDR, and
calls updateContentsHeadroom.

(WebCore::GPUCanvasContextCocoa::updateScreenHeadroomFromScreenProperties):
Fetches suppressHDR and current headroom from all screen properties, and
calls updateContentsHeadroom.

(WebCore::GPUCanvasContextCocoa::setDynamicRangeLimit):
WebGPU dynamic-range-limit handler, stores the new limits and:
- If there's no current screen data, calls
  updateScreenHeadroomFromScreenProperties, because dynamic-range-limit
  needs the current headroom and suppres-HDR flag to work,
- Otherwise just calls updateContentsHeadroom.

(WebCore::GPUCanvasContextCocoa::getEffectiveDynamicRangeLimitValue const):
For testing, combines dynamic-range-limit with suppress-HDR.

(WebCore::GPUCanvasContextCocoa::surfaceBufferToImageBuffer):
Refactored into updateScreenHeadroomFromScreenProperties.

(WebCore::GPUCanvasContextCocoa::configure):
Resets both current headroom and suppressHDR.

* Source/WebCore/rendering/RenderHTMLCanvas.h:
* Source/WebCore/rendering/RenderHTMLCanvas.cpp:
(WebCore::RenderHTMLCanvas::styleDidChange):
Catches dynamic-range-limit changes and notifies element.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::changeRequiresLayerRepaint const):
Ensures that dynamic-range-limit changes repaint the element, as its
effective colors may change.

* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::getContextEffectiveDynamicRangeLimitValue):
For testing: `internals.getContextEffectiveDynamicRangeLimitValue(canvas)`.

* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::screenPropertiesChanged):
Reworked suppress-headroom computation: Always 1.6, or current headroom
if lower.

* LayoutTests/http/tests/webgpu/webgpu/web_platform/canvas/dynamic-range-limit-constrained-disabled-expected.txt: Added.
* LayoutTests/http/tests/webgpu/webgpu/web_platform/canvas/dynamic-range-limit-constrained-disabled.html: Added.
* LayoutTests/http/tests/webgpu/webgpu/web_platform/canvas/dynamic-range-limit-expected.txt: Added.
* LayoutTests/http/tests/webgpu/webgpu/web_platform/canvas/dynamic-range-limit.html: Added.

Canonical link: https://commits.webkit.org/297105@main
